driver-core/src/main/com/mongodb/internal/connection/BaseCluster.java

```diff
@@ -326,14 +332,43 @@ private void logServerSelectionFailure(final ServerSelector serverSelector, fina
         }
     }
 
+    @Nullable
+    private ServerTuple selectServer(final ServerSelector serverSelector, final ClusterDescription clusterDescription) {
+        return selectServer(serverSelector, clusterDescription, this::getServer, getRandom());
+    }
 
-    private ServerTuple selectRandomServer(final ServerSelector serverSelector, final ClusterDescription clusterDescription) {
-        List<ServerDescription> serverDescriptions = serverSelector.select(clusterDescription);
-        if (!serverDescriptions.isEmpty()) {
-            return getRandomServer(new ArrayList<ServerDescription>(serverDescriptions));
-        } else {
-            return null;
+    @Nullable
+    @VisibleForTesting(otherwise = PRIVATE)
+    static ServerTuple selectServer(final ServerSelector serverSelector, final ClusterDescription clusterDescription,
+            final Function<ServerAddress, Server> serverCatalog, final Random random) {
+        return atMostNRandom(new ArrayList<>(serverSelector.select(clusterDescription)), 2, random, serverDescription -> {
+            Server server = serverCatalog.apply(serverDescription.getAddress());
+            return server == null ? null : new ServerTuple(server, serverDescription);
+        }).stream()
+                .min(comparingInt(serverTuple -> serverTuple.getServer().operationCount()))
+                .orElse(null);
+    }
+
+    /**
+     * Returns a new {@link List} of at most {@code n} elements, where each element is a result of
+     * {@linkplain Function#apply(Object) applying} the {@code transformer} to a randomly picked element from the specified {@code list},
+     * such that no element is picked more than once. If the {@code transformer} produces {@code null}, then another element is picked
+     * until either {@code n} transformed non-{@code null} elements are collected, or the {@code list} does not have
+     * unpicked elements left.
+     * <p>
+     * Note that this method may reorder the {@code list}.
+     */
+    private static <A, B, L extends List<A> & RandomAccess> List<B> atMostNRandom(final L list, final int n, final Random random,
```

Comment:
I would find this clearer without the generic types.  Certainly if this was a library function, this is the right way to do it, but seems overkill for a private method used in a single place.  We could always genericize it later if we need it elsewhere.  But if you want to leave it as is, I won't object.

Reply:
Done.